[Screen Time Refactoring] History fetching API does not work unless WKWebsiteDataStore has a delegate

jesxilin opened the pull request:
#### 6c635bbd1833196a956ee825b51445c775899749
<pre>
[Screen Time Refactoring] History fetching API does not work unless WKWebsiteDataStore has a delegate
<a href="https://bugs.webkit.org/show_bug.cgi?id=289553">https://bugs.webkit.org/show_bug.cgi?id=289553</a>
<a href="https://rdar.apple.com/145017247">rdar://145017247</a>

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Currently, the implemented getScreenTimeURLs does not get called when WKWebsiteDataStore does
not have a delegate. Instead, the virtual function for getScreenTimeURLs is called. Thus, the
current support for Screen Time in fetchDataRecordsOfTypes does not actually work.
As a consequence, since we never receive any records saying that they have Screen Time data,
removeDataOfTypes(forDataRecords specifically) also does not work.

Now, we add a ScreenTimeWebsiteDataSupport namespace that contains all the functions needed
for supporting Screen Time in conjunction with the WKWebsiteDataStore&apos;s fetchDataRecordsOfTypes
and removeDataOfTypes APIs. Due to these new files, WebColorPickerMac.mm has a build error where
it does not recognize its use of WeakPtr so include WeakPtr.

In addition, add an API test to ensure that the Screen Time&apos;s fetchAllHistoryWithCompletionHandler
API is being called. This tests fails with the old implementation.

Also, remove the old implementation from WebsiteDataStoreClient and replace all instances of
the old implementation with the working one.

Finally, in WebsiteDataStore.cpp, fix an assertion by making the completion handler
destructable on any thread. This is due to the problem where
STWebHistory.fetchAllHistoryWithCompletionHandler might not be called and just deallocates its
block. This can be removed once the fix to STWebHistory.fetchAllHistoryWithCompletionHandler
goes in.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h: Added.
* Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm: Added.
(WebKit::ScreenTimeWebsiteDataSupport::getScreenTimeURLs):
(WebKit::ScreenTimeWebsiteDataSupport::removeScreenTimeData):
(WebKit::ScreenTimeWebsiteDataSupport::removeScreenTimeDataWithInterval):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeScreenTimeData): Deleted.
(WebKit::WebsiteDataStore::removeScreenTimeDataWithInterval): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::getScreenTimeURLs const): Deleted.
* Source/WebKit/UIProcess/mac/WebColorPickerMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, FetchData)):

Canonical link: <a href="https://commits.webkit.org/292143@main">https://commits.webkit.org/292143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c127179d4782207dec832f379471759cb5c4160

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14694 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100129 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45600 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23116 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72536 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29824 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98100 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11188 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85859 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52861 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10896 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3593 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44938 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3721 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102176 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22141 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81537 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22388 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81884 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80929 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25494 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15383 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15267 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22114 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/27244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21772 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25246 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23512 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,94 @@
+// Copyright © 2025  All rights reserved.
```
this copyright comment should be the same as the rest of the files

aprotyas replied:
This is missing the proper copyright header

rr-codes commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,94 @@
+
+    // STWebHistory.fetchAllHistoryWithCompletionHandler sometimes deallocates its block instead of calling it.
+    // FIXME: Remove this once rdar://145889845 is widely available.
+    auto completionHandlerWithFinalizer = CompletionHandlerWithFinalizer<void(HashSet<URL>&&)>(WTFMove(completionHandler), [] (Function<void(HashSet<URL>&&)>& completionHandler) {
```
`Function<void(HashSet<URL>&&)>&` can be `auto&` I think

aprotyas replied:
Style guide nit, no space between capture list and arg list:

```suggestion
    auto completionHandlerWithFinalizer = CompletionHandlerWithFinalizer<void(HashSet<URL>&&)>(WTFMove(completionHandler), [](Function<void(HashSet<URL>&&)>& completionHandler) {
```

rr-codes commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,94 @@
+
+    if (!modifiedSince.isNaN()) {
+        NSTimeInterval timeInterval = modifiedSince.secondsSinceEpoch().seconds();
+        NSDate *date = [NSDate dateWithTimeIntervalSince1970:timeInterval];
```
NSDate * should be RetainPtr

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h:
```diff
@@ -0,0 +1,49 @@
+
+namespace WebKit {
+
+namespace ScreenTimeWebsiteDataSupport {
```
C++17 allows us to nest the namespaces!

```suggestion
namespace WebKit::ScreenTimeWebsiteDataSupport {
```

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h:
```diff
@@ -0,0 +1,49 @@
+
+#pragma once
+
+#include "WebsiteDataStoreConfiguration.h"
```
You can drop this include and forward declare `class WebsiteDataStoreConfiguration;`

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,94 @@
+        completionHandler({ });
+    });
+
+    [webHistory fetchAllHistoryWithCompletionHandler:makeBlockPtr([completionHandler = WTFMove(completionHandler)](NSSet<NSURL *> *urls, NSError *error) mutable {
```
Do you not mean to use `completionHandlerWithFinalizer`? `completionHandler` has already been moved in constructing its "with finalizer" version

jesxilin replied:
Good catch, yes I meant to use completionHandlerWithFinalizer

aprotyas replied:
```suggestion
    [webHistory fetchAllHistoryWithCompletionHandler:makeBlockPtr([completionHandler = WTFMove(completionHandlerWithFinalizer)](NSSet<NSURL *> *urls, NSError *error) mutable {
```

aprotyas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,94 @@
+void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDataStoreConfiguration& configuration)
+{
+    if (![PAL::getSTWebHistoryClass() instancesRespondToSelector:@selector(fetchAllHistoryWithCompletionHandler:)])
+        return;
```
Is this necessary? Since you're calling `-[STWebHistory deleteHistoryForURL:]` instead

aprotyas replied:
Ah, this isn't new code. You just moved this code in the PR. Feel free to keep this around

pxlcoder replied:
I think we should just remove it? No reason to keep it around.

jesxilin replied:
This was initially b/c if fetch doesn't work, then this would just get an empty vector as the dataRecords so it just early returns and does nothing. I can remove it.

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -546,6 +550,42 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+
+    TestWebKitAPI::Util::run(&calledScreenTimeFetchAllHistory);
+
+    EXPECT_TRUE(calledScreenTimeFetchAllHistory);
```
Unnecessary expectation? Since `run(&calledScreenTimeFetchAllHistory)` only returns once `calledScreenTimeFetchAllHistory == true`, right?

jesxilin replied:
Technically yes, but I just thought it was weird to not have an expectation in a test. I can take it out

aprotyas replied:
@jesxilin that's fair. Right now the test either passes or fails with a timeout.

You can keep your expectation so that it fails as an `EXPECT_` assertion.

If you do so, also set an upper bound on how long you want to wait for `calledScreenTimeFetchAllHistory` to flip (better to avoid timeout failures).

```
TestWebKitAPI::Util::runFor(&calledScreenTimeFetchAllHistory, 200_ms);
EXPECT_TRUE(calledScreenTimeFetchAllHistory);
```

jesxilin replied:
Ok, I'll use what you suggested. Thanks!

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h:
```diff
@@ -0,0 +1,48 @@
+
+class WebsiteDataStoreConfiguration;
+
+namespace ScreenTimeWebsiteDataSupport {
```
```suggestion
namespace ScreenTimeWebsiteDataSupport {

```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h:
```diff
@@ -0,0 +1,48 @@
+class WebsiteDataStoreConfiguration;
+
+namespace ScreenTimeWebsiteDataSupport {
+#if ENABLE(SCREEN_TIME)
```
Shouldn't the entire file be wrapped in  `#if ENABLE(SCREEN_TIME)`? (after the `#pragma once`)

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -546,6 +550,42 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+        })
+    };
+
+    RetainPtr dataTypeScreenTime = adoptNS([[NSSet alloc] initWithArray:@[_WKWebsiteDataTypeScreenTime]]);
```
Nit:
```suggestion
    RetainPtr dataTypeScreenTime = adoptNS([[NSSet alloc] initWithArray:@[ _WKWebsiteDataTypeScreenTime ]]);
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,116 @@
+#import <wtf/URL.h>
+#import <wtf/UUID.h>
+
+#if ENABLE(SCREEN_TIME)
```
Also think this file could be wrapped in `#if ENABLE(SCREEN_TIME)`.

Only these two lines could be before the `#if`.
```
#import "config.h"
#import "ScreenTimeWebsiteDataSupport.h"
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -91,6 +91,7 @@
 #endif
 
 #if PLATFORM(COCOA)
+#include "Cocoa/ScreenTimeWebsiteDataSupport.h"
```
```suggestion
#include "ScreenTimeWebsiteDataSupport.h"
```

jesxilin replied:
I added the Cocoa b/c EWS checks said they couldn't find the file. Is it actually safe to remove?

pxlcoder replied:
(yes, discussed offline)

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h:
```diff
@@ -0,0 +1,51 @@
+
+namespace WTF {
+class URL;
+class UUID;
```
Missing forward declarations for `HashSet`, `WallTime`, and `CompletionHandler`.

But I think you should just remove the WTF forward declarations and `#include <wtf/Forward.h>`.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -0,0 +1,111 @@
+#if ENABLE(SCREEN_TIME)
+
+#import "WebsiteDataStoreConfiguration.h"
+
```
```suggestion
#import "WebsiteDataStoreConfiguration.h"
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -39,6 +39,7 @@
 #include "NetworkProcessMessages.h"
 #include "PageLoadState.h"
 #include "RestrictedOpenerType.h"
+#include "ScreenTimeWebsiteDataSupport.h"
```
Move this to `#if PLATFORM(COCOA)`, after line 95. Will fix the GTK build failure.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -546,6 +550,42 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+        done = true;
+    }];
+
+    TestWebKitAPI::Util::run(&done);
```
You can get rid of this, and the `done` variable, now that you have `runFor` below.

pxlcoder replied:
Actually, this isn't entirely true, we still need to verify the completion handler gets called – not just that the swizzled method gets called.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -546,6 +550,42 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    InstanceMethodSwizzler swizzler {
+        PAL::getSTWebHistoryClass(),
+        @selector(fetchAllHistoryWithCompletionHandler:),
+        imp_implementationWithBlock(^(id object, NSSet<NSURL *> *urls, NSError *error) {
```
Couple things:
1. This method signature is incorrect, since `fetchAllHistoryWithCompletionHandler:` take a single completion handler parameter. Not a set and an error parameter.
2. After fixing that, we can call the completion handler with a mock set of URLs to achieve better testing.
3. We don't need the `runFor` with a timeout, since we're not actually calling into the XPC service.

jesxilin committed:
[Screen Time Refactoring] History fetching API does not work unless WKWebsiteDataStore has a delegate
https://bugs.webkit.org/show_bug.cgi?id=289553
rdar://145017247

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Currently, the implemented getScreenTimeURLs does not get called when WKWebsiteDataStore does
not have a delegate. Instead, the virtual function for getScreenTimeURLs is called. Thus, the
current support for Screen Time in fetchDataRecordsOfTypes does not actually work.
As a consequence, since we never receive any records saying that they have Screen Time data,
removeDataOfTypes(forDataRecords specifically) also does not work.

Now, we add a ScreenTimeWebsiteDataSupport namespace that contains all the functions needed
for supporting Screen Time in conjunction with the WKWebsiteDataStore's fetchDataRecordsOfTypes
and removeDataOfTypes APIs. Due to these new files, WebColorPickerMac.mm has a build error where
it does not recognize its use of WeakPtr so include WeakPtr.

In addition, add an API test to ensure that the Screen Time's fetchAllHistoryWithCompletionHandler
API is being called. This tests fails with the old implementation.

Also, remove the old implementation from WebsiteDataStoreClient and replace all instances of
the old implementation with the working one.

Finally, in WebsiteDataStore.cpp, fix an assertion by making the completion handler
destructable on any thread. This is due to the problem where
STWebHistory.fetchAllHistoryWithCompletionHandler might not be called and just deallocates its
block. This can be removed once the fix to STWebHistory.fetchAllHistoryWithCompletionHandler
goes in.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.h: Added.
* Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm: Added.
(WebKit::ScreenTimeWebsiteDataSupport::getScreenTimeURLs):
(WebKit::ScreenTimeWebsiteDataSupport::removeScreenTimeData):
(WebKit::ScreenTimeWebsiteDataSupport::removeScreenTimeDataWithInterval):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeScreenTimeData): Deleted.
(WebKit::WebsiteDataStore::removeScreenTimeDataWithInterval): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::getScreenTimeURLs const): Deleted.
* Source/WebKit/UIProcess/mac/WebColorPickerMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, FetchData)):

Canonical link: https://commits.webkit.org/292143@main
